[BE][Docs] Improve dist.new_group doc

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55660 [BE][Docs] Improve dist.new_group doc**

Noticed this doc was missing clarification on nccl env vars that
init_process_group docs have. Also, specify default behavior when backend=None
is passed in.

Differential Revision: [D27672208](https://our.internmc.facebook.com/intern/diff/D27672208/)

rohan-varma committed:
[BE][Docs] Improve dist.new_group doc

Noticed this doc was missing clarification on nccl env vars that
init_process_group docs have. Also, specify default behavior when backend=None
is passed in.

Differential Revision: [D27672208](https://our.internmc.facebook.com/intern/diff/D27672208/)

[ghstack-poisoned]

SciPioneer commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2569,13 +2569,30 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+            done since CUDA execution is async and it is no longer safe to
+            continue executing user code since failed async NCCL operations
+            might result in subsequent CUDA operations running on corrupted
+            data. Only one of these two environment variables should be set.
```
Can we add a sanity check in DDP constructor to make sure at most only one of these two env variables is set?

rohan-varma replied:
I think something like this is already done in `ProcessGroupNCCL` - https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/ProcessGroupNCCL.cpp#L431. We use blocking wait if both are enabled.

rohan-varma committed:
Update on "[BE][Docs] Improve dist.new_group doc"


Noticed this doc was missing clarification on nccl env vars that
init_process_group docs have. Also, specify default behavior when backend=None
is passed in.

Differential Revision: [D27672208](https://our.internmc.facebook.com/intern/diff/D27672208/)

[ghstack-poisoned]
